File: backend/logger.py
import os
import json
from datetime import datetime

# Base path setup
BACKEND_DIR = os.path.dirname(os.path.abspath(__file__))
LOGS_PATH = os.path.join(BACKEND_DIR, 'logs.json')

MAX_LOG_ENTRIES = 1000

DEFAULT_LOGS = {
    "logs": []
}

def load_logs():
    """Load logs from logs.json. Automatically initializes if missing or invalid."""
    if not os.path.exists(LOGS_PATH):
        save_logs(DEFAULT_LOGS)
        return DEFAULT_LOGS
    
    try:
        with open(LOGS_PATH, 'r', encoding='utf-8') as f:
            data = json.load(f)
            if not isinstance(data, dict) or "logs" not in data or not isinstance(data["logs"], list):
                raise ValueError("Logs root is invalid")
            return data
    except Exception:
        # Recreate logs if malformed
        save_logs(DEFAULT_LOGS)
        return DEFAULT_LOGS

def save_logs(logs_data):
    """Save logs to logs.json."""
    try:
        os.makedirs(os.path.dirname(LOGS_PATH), exist_ok=True)
        with open(LOGS_PATH, 'w', encoding='utf-8') as f:
            json.dump(logs_data, f, indent=4)
        return True
    except Exception:
        return False

def log_event(project, status, message, stdout="", stderr=""):
    """
    Log an event to logs.json. Caps total logs to MAX_LOG_ENTRIES.
    Statuses supported: SUCCESS, FAILED, NO_CHANGES, WARNING, PENDING_RETRY, PAUSED, DISABLED, ALREADY_RUNNING
    """
    logs_data = load_logs()
    
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    
    new_entry = {
        "timestamp": timestamp,
        "project": project if project else "System",
        "status": status,
        "message": message,
        "stdout": str(stdout) if stdout is not None else "",
        "stderr": str(stderr) if stderr is not None else ""
    }
    
    logs_data["logs"].insert(0, new_entry)  # Prepend so newest is at the top
    
    # Cap size
    if len(logs_data["logs"]) > MAX_LOG_ENTRIES:
        logs_data["logs"] = logs_data["logs"][:MAX_LOG_ENTRIES]
        
    save_logs(logs_data)
    return new_entry

def get_logs():
    """Get all logs (newest first)."""
    logs_data = load_logs()
    return logs_data.get("logs", [])

File: src/main/store/logger.ts
import { existsSync, readFileSync } from 'fs'
import type { LogEntry } from '../../shared/types'
import { atomicWriteJson } from './atomic'
import { logsPath } from './paths'

const MAX_LOG_ENTRIES = 1000

interface LogsFile {
  logs: LogEntry[]
}

const DEFAULT_LOGS: LogsFile = { logs: [] }

/** "YYYY-MM-DD HH:MM:SS" in local time, matching logger.py. */
export function formatTimestamp(d = new Date()): string {
  const pad = (n: number): string => String(n).padStart(2, '0')
  return (
    `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ` +
    `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`
  )
}

export function loadLogs(): LogsFile {
  const path = logsPath()
  if (!existsSync(path)) {
    saveLogs(DEFAULT_LOGS)
    return { logs: [] }
  }
  try {
    const data = JSON.parse(readFileSync(path, 'utf-8')) as unknown
    if (
      typeof data !== 'object' ||
      data === null ||
      !Array.isArray((data as LogsFile).logs)
    ) {
      throw new Error('Logs root is invalid')
    }
    return data as LogsFile
  } catch {
    saveLogs(DEFAULT_LOGS)
    return { logs: [] }
  }
}

export function saveLogs(logs: LogsFile): boolean {
  try {
    atomicWriteJson(logsPath(), logs)
    return true
  } catch {
    return false
  }
}

/** Prepend a log entry (newest first) and cap at MAX_LOG_ENTRIES. */
export function logEvent(
  project: string,
  status: string,
  message: string,
  stdout: string = '',
  stderr: string = ''
): LogEntry {
  const data = loadLogs()
  const entry: LogEntry = {
    timestamp: formatTimestamp(),
    project: project || 'System',
    status,
    message,
    stdout: stdout != null ? String(stdout) : '',
    stderr: stderr != null ? String(stderr) : ''
  }
  data.logs.unshift(entry)
  if (data.logs.length > MAX_LOG_ENTRIES) {
    data.logs = data.logs.slice(0, MAX_LOG_ENTRIES)
  }
  saveLogs(data)
  return entry
}

export function getLogs(): LogEntry[] {
  return loadLogs().logs
}
